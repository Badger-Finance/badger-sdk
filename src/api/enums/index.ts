export { Currency } from './currency.enum';
