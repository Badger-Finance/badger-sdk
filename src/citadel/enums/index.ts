export * from './reward-data-type.enum';
export * from './reward-event-type.enum';
export * from './reward-filter.enum';
