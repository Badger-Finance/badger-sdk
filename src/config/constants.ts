import { ARB_CONFIG } from './network/arbitrum.config';
import { AVAX_CONFIG } from './network/avalanche.config';
import { BSC_CONFIG } from './network/binance.config';
import { ETH_CONFIG } from './network/ethereum.config';
import { FTM_CONFIG } from './network/fantom.config';
import { LOCAL_CONFIG } from './network/local.config';
import { MATIC_CONFIG } from './network/polygon.config';

export const ONE_MIN_MS = 60 * 1000;
export const ONE_HOUR_MS = 60 * ONE_MIN_MS;
export const ONE_DAY_MS = 24 * ONE_HOUR_MS;
export const ONE_YEAR_MS = 365 * ONE_DAY_MS;

export const SUPPORTED_NETWORKS = [
  ETH_CONFIG,
  LOCAL_CONFIG,
  BSC_CONFIG,
  MATIC_CONFIG,
  ARB_CONFIG,
  AVAX_CONFIG,
  FTM_CONFIG,
];

export const VERSION = '1.0.77';
