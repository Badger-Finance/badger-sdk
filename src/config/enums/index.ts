export { Network } from './network.enum';
