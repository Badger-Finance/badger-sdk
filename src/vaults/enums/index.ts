export { VaultVersion } from './vault-version.enum';
