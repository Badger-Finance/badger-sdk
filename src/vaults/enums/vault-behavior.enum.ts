export enum VaultBehavior {
  Compounder = 'Compounder',
  DCA = 'DCA',
  Ecosystem = 'Ecosystem',
  EcosystemHelper = 'Ecosystem Helper',
  None = 'None',
}
