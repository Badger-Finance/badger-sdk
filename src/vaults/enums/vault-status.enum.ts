export enum VaultStatus {
  deprecated,
  experimental,
  guarded,
  open,
}
