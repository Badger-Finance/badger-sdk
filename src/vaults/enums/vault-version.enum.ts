export enum VaultVersion {
  v1 = 'v1',
  v1_5 = 'v1.5',
  v2 = 'v2',
}
