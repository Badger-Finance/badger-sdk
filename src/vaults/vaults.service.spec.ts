import { BigNumber } from '@ethersproject/bignumber';
import { any, MockProxy } from 'jest-mock-extended';

import { Controller, StrategyV15, TokensService, Vault, VaultV15 } from '..';
import { BaseStrategy } from '../contracts/StrategyV15';
import { BadgerSDK } from '../sdk';
import { TEST_ADDR } from '../tests/tests.constants';
import { mockSDK, mockVaults } from '../tests/tests.utils';
import { VaultVersion } from './enums';

describe('vaults.service', () => {
  let sdk: BadgerSDK;
  let controller: MockProxy<Controller>;
  let vault: MockProxy<Vault>;
  let vaultV15: MockProxy<VaultV15>;
  let strategyV15: MockProxy<StrategyV15>;

  beforeEach(async () => {
    sdk = mockSDK();

    // destructure assignment from mocks
    ({ controller, vault, vaultV15, strategyV15 } = mockVaults());

    // Let the SDK Prepare
    await sdk.ready();
  });

  describe('getPendingYield', () => {
    it('returns pending reward tokens from the underlying yield source', async () => {
      vaultV15.strategy.calledWith().mockImplementation(async () => TEST_ADDR);
      vaultV15.lastHarvestedAt
        .calledWith()
        .mockImplementation(async () => BigNumber.from('1649865952'));

      const rewards = [
        {
          token: '0xD533a949740bb3306d119CC777fa900bA034cd52',
          amount: BigNumber.from('3140000000000000000'),
        }, // 3.14 CRV
        {
          token: '0x4e3fbd56cd56c3e72c1403e103b45db9da5b9d2b',
          amount: BigNumber.from('69000000000000000'),
        }, // 0.69 CVX
      ] as BaseStrategy.TokenAmountStructOutput[];
      strategyV15.balanceOfRewards
        .calledWith(any())
        .mockImplementation(async () => rewards);

      jest
        .spyOn(TokensService.prototype, 'loadToken')
        .mockImplementation(async (token) => {
          if (token === '0xD533a949740bb3306d119CC777fa900bA034cd52') {
            // CRV
            return {
              address: '0xD533a949740bb3306d119CC777fa900bA034cd52',
              name: 'Curve DAO Token',
              symbol: 'CRV',
              decimals: 18,
            };
          }
          return {
            address: '0x4e3fbd56cd56c3e72c1403e103b45db9da5b9d2b',
            name: 'Convex DAO Token',
            symbol: 'CVX',
            decimals: 18,
          };
        });

      const actual = await sdk.vaults.getPendingYield(TEST_ADDR);
      expect(actual).toMatchSnapshot();
    });
  });

  describe('getVaultStrategy', () => {
    it('should return v1.5 strategy', async function () {
      // BVEOXD vault that is v1.5 vault
      const address = '0x96d4dBdc91Bef716eb407e415c9987a9fAfb8906';
      const strategy = '0x0c7E0807011A218d0F1A156D3965875ff233933E';
      vaultV15.strategy.calledWith().mockImplementation(async () => strategy);
      const version = VaultVersion.v1_5;
      const strategyAddress = await sdk.vaults.getVaultStrategy({
        address,
        version,
      });
      expect(strategyAddress).toEqual(strategy);
    });
    it('should return v1 strategy', async function () {
      // BSMM_WBTC_RENBTC vault that is v1 vault
      const address = '0xb6d63a4e5ca740e96c26adabcac73be78ee39dc5';
      const strategy = '0x711555f2B421DA9A86a18Dc163d04699310fE297';
      vault.controller.calledWith().mockImplementation(async () => TEST_ADDR);
      vault.token.calledWith().mockImplementation(async () => TEST_ADDR);
      controller.strategies
        .calledWith(any())
        .mockImplementation(async () => strategy);
      const version = VaultVersion.v1;
      const strategyAddress = await sdk.vaults.getVaultStrategy({
        address,
        version,
      });
      expect(strategyAddress).toEqual(strategy);
    });
    it('incompatible vault should throw', async function () {
      const address = '0x96d4dBdc91Bef716eb407e415c9987a9fAfb8906';
      vault.controller.calledWith().mockImplementation(async () => {
        throw new Error('v1.5 Vault Requested!');
      });
      // Set incompatible vault version
      const version = VaultVersion.v1;
      await expect(
        async () => await sdk.vaults.getVaultStrategy({ address, version }),
      ).rejects.toThrow(Error);
    });
  });
});
