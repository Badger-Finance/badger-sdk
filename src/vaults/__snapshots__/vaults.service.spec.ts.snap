// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vaults.service getPendingYield returns pending reward tokens from the underlying yield source 1`] = `
Object {
  "lastHarvestedAt": 1649865952,
  "tokenRewards": Array [
    Object {
      "address": "0xD533a949740bb3306d119CC777fa900bA034cd52",
      "balance": 3.14,
      "decimals": 18,
      "name": "Curve DAO Token",
      "symbol": "CRV",
    },
    Object {
      "address": "0x4e3fbd56cd56c3e72c1403e103b45db9da5b9d2b",
      "balance": 0.069,
      "decimals": 18,
      "name": "Convex DAO Token",
      "symbol": "CVX",
    },
  ],
}
`;
